relay-server/src/actors/processor.rs

```diff
@@ -1848,34 +1839,42 @@ impl EnvelopeProcessor {
                 }
             },
             || {
-                let envelope_context = state.envelope_context;
+                let project_key = state.envelope_context.scoping().project_key;
 
                 match self.process_state(&mut state) {
                     Ok(()) => {
                         if !state.extracted_metrics.is_empty() {
                             let project_cache = ProjectCache::from_registry();
-                            project_cache.do_send(InsertMetrics::new(
-                                envelope_context.scoping().project_key,
-                                state.extracted_metrics,
-                            ));
+                            project_cache
+                                .do_send(InsertMetrics::new(project_key, state.extracted_metrics));
                         }
 
+                        // The envelope could be modified or even emptied during processing, which
+                        // requires recomputation of the context.
+                        state.envelope_context.update(&state.envelope);
+
+                        let envelope_response = if state.envelope.is_empty() {
+                            // Individual rate limits have already been issued
+                            state.envelope_context.reject(Outcome::RateLimited(None));
+                            None
+                        } else {
+                            Some((state.envelope, state.envelope_context))
+                        };
```

Comment:
Same logic in the the processor (ProcessEnvelope) and the project (CheckEnvelope): 

 1. Update the context with new data from the modified envelope. Particularly rate limits change the composition of items which has an effect on outcomes.
 2. Check if the envelope is now empty and drop it potentially. The call to `reject()` is not strictly necessary. The passed outcome will be ignored since the envelope is empty, and simply rejection metrics will be logged. However, it's better to be explicit for future-proofing. 
 3. If there are still items in the envelope, return it along with the updated context.